### ISSUE TITLE
ProjectCodes with lowercase letters: Token missing required scope to update project

### ISSUE THREAD

**Issue by benjaminran** [BD2KGenomics/dcc-storage#26]
A valid access token is denied authorization or upload to download files whose project the token does have scope for if the project name contains lowercase letters:

```
$ sudo env "PATH=$PATH" redwood token create -u 'testuser1' -s 'aws.Quake_Brain_scRNA-Seq.upload aws.Quake_Brain_scRNA-Seq.download'
XXXXXXXXXXXXXXX

~~~

$ sudo env "PATH=$PATH" redwood token list -u 'testuser1' | jq
{
  "tokens": [
    {
      "desc": "test access token",
      "scope": [
        "aws.Quake_Brain_scRNA-Seq.download",
        "aws.Quake_Brain_scRNA-Seq.upload"
      ],
      "access_token": "XXXXXXXXXXXXXX",
      "expires_in": 31535987
    }
  ]
}

~~~

<in redwood-client:1.1.1>
$ export ACCESS_TOKEN=XXXXXXXXXXX
$ upload -p 'Quake_Brain_scRNA-Seq' ./foo.txt
Reading /tmp/tmp.H0cMg0Z75X/registration.txt 
[1/1] Registered ./foo.txt
ERROR: Command error: java.io.IOException: Storage client error: Token missing required scope to update project

Please check the log for detailed error messages

~~~

icgc-storage-client.log:
========================
2017-06-06 23:26:17,660 [main] INFO  o.i.d.s.c.c.UploadCommand - Uploading file './foo.txt'...
2017-06-06 23:26:17,701 [main] WARN  o.i.d.s.c.e.ServiceRetryableResponseErrorHandler - Access Denied. Stop processing: 
2017-06-06 23:26:17,706 [main] ERROR o.i.d.s.c.ClientMain - Unknown error: 
org.icgc.dcc.storage.client.exception.NotResumableException: java.io.IOException: Storage client error: Token missing required scope to update project
	at org.icgc.dcc.storage.client.exception.ControlExceptionFactory.notResumableException(ControlExceptionFactory.java:54)
	at org.icgc.dcc.storage.client.exception.ServiceRetryableResponseErrorHandler.handleError(ServiceRetryableResponseErrorHandler.java:55)
	at org.springframework.web.client.RestTemplate.handleResponse(RestTemplate.java:667)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:620)
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:580)
	at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:498)
	at org.icgc.dcc.storage.client.transport.StorageService.lambda$isObjectExist$5(StorageService.java:263)
	at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:263)
	at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:154)
	at org.icgc.dcc.storage.client.transport.StorageService.isObjectExist(StorageService.java:262)
	at org.icgc.dcc.storage.client.upload.s3.S3UploadService.isObjectExist(S3UploadService.java:233)
	at org.icgc.dcc.storage.client.command.UploadCommand.uploadFile(UploadCommand.java:106)
	at org.icgc.dcc.storage.client.command.UploadCommand.execute(UploadCommand.java:90)
	at org.icgc.dcc.storage.client.ClientMain.run(ClientMain.java:136)
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:800)
	at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:784)
	at org.springframework.boot.SpringApplication.afterRefresh(SpringApplication.java:771)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:316)
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:134)
	at org.icgc.dcc.storage.client.ClientMain.main(ClientMain.java:109)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:58)
Caused by: java.io.IOException: Storage client error: Token missing required scope to update project
	... 28 common frames omitted
2017-06-06 23:26:17,712 [Thread-1] INFO  o.s.c.a.AnnotationConfigApplicationContext - Closing org.springframework.context.annotation.AnnotationConfigApplicationContext@13969fbe: startup date [Tue Jun 06 23:26:15 UTC 2017]; root of context hierarchy

~~~

storage-server log:
===================
2017-06-06 23:26:17,674 [http-nio-5431-exec-5] INFO  o.i.d.s.s.s.AbstractScopeAuthorizationStrategy - Checking authorization with object id 80abe89e-1c83-5776-858c-e0ff9b7df128
2017-06-06 23:26:17,677 [http-nio-5431-exec-5] INFO  o.i.d.s.s.s.ProjectScopeStrategy - checking for permission to project Quake_Brain_scRNA-Seq for object id 80abe89e-1c83-5776-858c-e0ff9b7df128 (false)

~~~

$ curl -k https://metadata.ucsc-cgp.org/entities/80abe89e-1c83-5776-858c-e0ff9b7df128 | jq
{
  "id": "80abe89e-1c83-5776-858c-e0ff9b7df128",
  "gnosId": "2daec1a3-0170-42e2-aaaa-abc1164a58dd",
  "fileName": "foo.txt",
  "projectCode": "Quake_Brain_scRNA-Seq",
  "access": "controlled",
  "createdTime": 1496791573775
}

~~~

<in redwood-client:1.1.1>
$ cat registration.tsv
gnos_id	program_code	filename	md5	access
2daec1a3-0170-42e2-aaaa-abc1164a58dd	Quake_Brain_scRNA-Seq	./foo.txt	dfa1b0b6ef47a15c600b094a93e56002	controlled

~~~

$ curl -k -XPOST https://auth.ucsc-cgp.org/oauth/check_token -d token=XXXXXXXXXXXXXXXX -u mgmt:XXXX
{
  "exp": 1528327525,
  "user_name": "testuser1",
  "client_id": "mgmt",
  "scope": [
    "aws.Quake_Brain_scRNA-Seq.download",
    "aws.Quake_Brain_scRNA-Seq.upload"
  ],
  "desc": "test access token"
}
```

TLDR; Token and metadata entity have the same scope (project component) and projectCode respectively, but access is denied anyway.

The problem is that the storage-server converts the auth token scopes into [AuthToken](https://github.com/BD2KGenomics/dcc-storage/blob/369b8b9e7b06d420192d732f39d573d0f28f34d7/dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/security/AuthScope.java) objects, which uppercases the project component of the scope. Those uppercased projects from the token scopes are checked against the [unmodified](https://github.com/BD2KGenomics/dcc-storage/blob/369b8b9e7b06d420192d732f39d573d0f28f34d7/dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/security/ProjectScopeStrategy.java#L107) (i.e. non-uppercased) programCode from the metadata Entity for the file being requested.

I think the storage-server shouldn't modify the tokens after pulling them out from the database because then the exact values from the auth-db and metadata-db will be being compared--no transformations to have to keep track of. A nearly-as-good solution would be to enforce the uppercasing of the metadata projectCodes; that would fix this issue but would still require us to remember that this transformation happens, especially since the metadata-server and auth-server don't enforce an all-caps requirement when registering bundles or creating auth-tokens

**Comment by benjaminran** [comment]
Additional note: redwood projectCodes cannot contain periods (that'll trip up the way the storage-server parses them into site.project.action).
